### PR TITLE
:book: Add Roadmaps to bring visibility and allow better collaboration

### PR DESCRIPTION
**Pull request by camilamacedo86**
## Overview

This PR introduces the addition of detailed roadmaps to our project. These documents outline our strategic objectives, planned initiatives, and focus areas for the upcoming periods. The primary goal is to enhance visibility into the project's direction and foster a more collaborative environment for both current contributors and newcomers.

## What's Included

- **Project Roadmap 2024**: Outlines our major goals and initiatives for 2024, focusing on improving maintainability, user experience, and integrating new functionalities that meet our community's needs.

## Goals

- **Enhance Visibility**: By clearly articulating our plans and priorities, we aim to provide all stakeholders with a transparent view of where the project is headed.
- **Encourage Collaboration**: The roadmaps are designed to invite contributions and feedback from the community, ensuring that the project benefits from a diverse range of perspectives and expertise.
- **Streamline Contribution Process**: New and existing contributors can easily identify areas where their efforts can have the most impact, aligning with the project's strategic goals.

Thank you for your support and collaboration!

### CODE REVIEW

**Comment by camilamacedo86** [comment]
/hold

Please folks feel free to LGTM and approve this one
We are holding to let more people review.

File: roadmap/roadmap_2024.md
Comment (Kavinjsir) [inline roadmap/roadmap_2024.md:16]: nit: we can add `---` to separate each project for better readability.

File: roadmap/roadmap_2024.md
Comment (Kavinjsir) [inline roadmap/roadmap_2024.md:93]: Some topics have come to mind from the previous discussions in the community:

### 1\. Server Side Apply (SSA)

This issue has a [long history](https://kubernetes.slack.com/archives/CAR30FCJZ/p1645534614215639) and appears to be [blocked](https://github.com/kubernetes-sigs/kubebuilder/issues/2514) due to challenges with upstream components. However, the controller-runtime team seems to be actively working on `SSA` recently. See:

1.  [kubernetes/kubernetes#118138](https://github.com/kubernetes/kubernetes/issues/118138)
2.  [controller-runtime#2702](https://github.com/kubernetes-sigs/controller-runtime/pull/2702)

Perhaps we'll have the opportunity to address this within the year?

### 2\. Updating Controller Scaffolding to Align with the Latest Interface from controller-runtime

Since `v3` of Kubebuilder (kb), we've provided a stable scaffolding layout to facilitate controller development.
However, some interfaces from `controller-runtime` appear outdated, and recent versions have introduced new structures/interfaces.
I'm wondering if it's time to reevaluate our scaffolding code template and consider trimming the use of legacy structures/interfaces if necessary?

**Review by Kavinjsir — APPROVED** [review]
Super excited about the roadmap and very much looking forward to engaging in contributions based on the expected trajectory!

Just left a nit and raised a proposal for some topics previously discussed in the community.

I also have some questions:

1.  How do we plan to maintain the roadmap each year? I think it might be necessary for us to evaluate it periodically, which could help us check if we are heading in the right direction and if new topics of interest are emerging from the community.
2.  I like the structure of each topic, which includes `status`, `objective`, `motivation`, and `context`. Would it be possible to create a template for this so that every topic can be aligned in the same format?

File: roadmap/roadmap_2024.md
Comment (camilamacedo86) [reply roadmap/roadmap_2024.md:16]: Done

File: roadmap/roadmap_2024.md
Comment (camilamacedo86) [reply roadmap/roadmap_2024.md:90]: Good points. I added the item 2 but not 1.
The `Server Side Apply (SSA)` for now still an objetive of Controller-Tools
We only will be able to have it as objetive for us when they implement that

Also, I think we need to try to keep the list small
Not more than 5 goals in the year.
We have no available effort to looking for all. But I either agree that the most import goals needs to be there.

**Review by camilamacedo86 — COMMENTED** [review]
Hi @Kavinjsir 

I tried to address all your comments. Following the answer of your questions inline

>How do we plan to maintain the roadmap each year? I think it might be necessary for us to evaluate it periodically, which could help us check if we are heading in the right direction and if new topics of interest are emerging from the community.

If we see that we need to add an macro goal we can push a PR within and to do the discussion. However, we need to keep in mind that we have not a lot of effort available
So, I would say for we try to keep the list short with 5 items in the max. Just what is more
important/crucial in the year.

However, yes. That is live doc and we can and should keep it update when possible.
For example, if we delivery something we should update to complete so that we know
that goal was accomplished in the year. 

Nothing is write in a stone :-) 
We can either remove/change and etc any goal if we see that is required. 
The idea is to bring more visibility about what we need / what we want to acomplish
the direction of the project and see if we can win more help as well since it makes clear what is most important at the moment. 

> I like the structure of each topic, which includes status, objective, motivation, and context. Would it be possible to create a template for this so that every topic can be aligned in the same format?

It was added. Good Point 🥇

File: roadmap/roadmap_2024.md
Comment (Kavinjsir) [reply roadmap/roadmap_2024.md]: Glad to hear that the information on `SSA` and the annual goals was helpful. Thanks for the clarification!

**Comment by Kavinjsir** [comment]
> Hi @Kavinjsir
> 
> I tried to address all your comments. Following the answer of your questions inline
> 
> > How do we plan to maintain the roadmap each year? I think it might be necessary for us to evaluate it periodically, which could help us check if we are heading in the right direction and if new topics of interest are emerging from the community.
> 
> If we see that we need to add an macro goal we can push a PR within and to do the discussion. However, we need to keep in mind that we have not a lot of effort available So, I would say for we try to keep the list short with 5 items in the max. Just what is more important/crucial in the year.
> 
> However, yes. That is live doc and we can and should keep it update when possible. For example, if we delivery something we should update to complete so that we know that goal was accomplished in the year.
> 
> Nothing is write in a stone :-) We can either remove/change and etc any goal if we see that is required. The idea is to bring more visibility about what we need / what we want to acomplish the direction of the project and see if we can win more help as well since it makes clear what is most important at the moment.
> 
> > I like the structure of each topic, which includes status, objective, motivation, and context. Would it be possible to create a template for this so that every topic can be aligned in the same format?
> 
> It was added. Good Point 🥇

Hey @camilamacedo86, I appreciate your comment!
It totally makes sense to keep it a live document with the flexibility to incorporate community feedback.

The way I see the proposed roadmap is as a guide to ensure we're moving Kubebuilder (kb) in the right direction.
Personally, I'd like to suggest evaluating the docs periodically, perhaps quarterly through meetings.

Additionally, I wonder if it's possible or necessary for us to align our goals with KubeCon events.
This could help us disseminate updates about Kubebuilder and gather more feedback from a wider audience.

File: roadmap/roadmap_2024.md
Comment (camilamacedo86) [reply roadmap/roadmap_2024.md]: Closing this one but feel free to re-open
REALLY THANK YOU FOR ALL YOUR SUPPORT AND COLLAB !!

**Comment by camilamacedo86** [comment]
Hi @Kavinjsir 

> Additionally, I wonder if it's possible or necessary for us to align our goals with KubeCon events.

About this one, if we do any other talk about Kubebuilder in the contribution track for example I think we can share it for sure and let me people comment if they wish to do so.

**Comment by camilamacedo86** [comment]
/override pull-kubebuilder-e2e-k8s-1-29-2

Reason:https://kubernetes.slack.com/archives/CCK68P2Q2/p1712378775102839

File: roadmap/roadmap_2024.md
Comment (camilamacedo86) [inline roadmap/roadmap_2024.md:16]:
```suggestion
###  kube-rbac-proxy's Role in Default Scaffold
```

File: roadmap/roadmap_2024.md
Comment (camilamacedo86) [inline roadmap/roadmap_2024.md:18]:
```suggestion
**Status:** 🚧 : WIP: Open for Discussion/Proposal Pending; Contributions Welcome. [More info](https://github.com/kubernetes-sigs/kubebuilder/pull/3853)
```

**Comment by camilamacedo86** [comment]
Hi @varshaprasad96 

I think this one is good to fly.
The idea is change it and keep it updated as well

So, we can always address follow ups.
WDYT?
Could we get this one merged?

**Comment by camilamacedo86** [comment]
/hold cancel

We shared it in the malling list + channel
We gave a good time for objections and reviews
We can change things up always 
So, I am removing the hold to allow it get merged within the next `lgtm`

File: roadmap/roadmap_2024.md
Comment (varshaprasad96) [inline roadmap/roadmap_2024.md:90]: We could also add that this involves updating dependencies based on k8s bumps. It would be nice to have community's help on this too.

**Review by varshaprasad96 — APPROVED** [review]
Thanks @camilamacedo86! This looks good! Just one nit on adding about k8s bumps to keep KB in sync with recent changes in the community, but that is indirectly implied from controller-runtime's bump.

/lgtm
/approve

File: roadmap/roadmap_2024.md
Comment (camilamacedo86) [reply roadmap/roadmap_2024.md:90]: I have no objetions about we add a follow up like "Keep the project updated with the edge" so that we clarify that is a goal as well for all deps.

However, I think we have this goal already by default right. It might be implicit in keep the project maintained.

**Pull request merged**
